src/Microsoft.ApiDesignGuidelines.Analyzers/Core/TypeNamesShouldNotMatchNamespaces.cs

```diff
@@ -25,24 +29,236 @@ public abstract class TypeNamesShouldNotMatchNamespacesAnalyzer : DiagnosticAnal
         internal static DiagnosticDescriptor SystemRule = new DiagnosticDescriptor(RuleId,
                                                                              s_localizableTitle,
                                                                              s_localizableMessageSystem,
                                                                              DiagnosticCategory.Naming,
                                                                              DiagnosticSeverity.Warning,
-                                                                             isEnabledByDefault: false,
+                                                                             isEnabledByDefault: true,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182257.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule, SystemRule);
 
+        private static object s_obj = new object();
+        private static ImmutableDictionary<string, string> s_wellKnownSystemNamespaceTable;
+
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterCompilationStartAction(
+                compilationStartAnalysisContext =>
+                {
+                    var namedTypesInCompilation = new ConcurrentBag<INamedTypeSymbol>();
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        SymbolAnalysisContext =>
+                        {
+                            var namedType = (INamedTypeSymbol)SymbolAnalysisContext.Symbol;
+                            namedTypesInCompilation.Add(namedType);
+                        }, SymbolKind.NamedType);
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        compilationAnalysisContext =>
+                        {
+                            var namespaceNamesInCompilation = new ConcurrentBag<string>();
+                            var compilation = compilationAnalysisContext.Compilation;
+                            AddNamespacesFromCompilation(namespaceNamesInCompilation, compilation.GlobalNamespace);
+
+                            // The entries are added in the sorted order so that the diagnostic message is always deterministic
+                            var namespaceTableForCompilation = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```

Comment:
The idea is right but there is a small change here.

When we declare namespace A.B and C.D. We have 4 namespaces. Namely, "A", "A.B", "C" and "C.D". So the mapping will look like "A" to "A", "B" to "A.B", "C" to "C", and "D" to "C.D".

A suitable example to explain the sorting would be to use the namespaces "A" and "A.B" and expect a diagnostic "Type name "A" conflicts with namespace name "A"(and not "A.B")" all the time